From b32ce9509d4039e56f527e2874550eb2cf1a5648 Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Thu, 9 Jan 2014 11:50:51 +0100
Subject: [PATCH 1/1] uvm: add support for Linux 3.12 (v2)

If CONFIG_UIDGID_STRICT_TYPE_CHECKS is enabled, kuid_t is a struct
with the uid_t val member, otherwise it's simply an alias for uid_t.

Despite the commit subject, this is not specific to Linux 3.12.

Thanks to Joseph Yasi for finding out the issue and suggesting a fix.
---
 nvidia_uvm_linux.h |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/nvidia_uvm_linux.h b/nvidia_uvm_linux.h
index e2b4620..0a38cbc 100644
--- a/nvidia_uvm_linux.h
+++ b/nvidia_uvm_linux.h
@@ -389,11 +389,17 @@ typedef void irqreturn_t;
 // not require the RCU's read lock on current->cred.
 //
 //
+#ifdef CONFIG_UIDGID_STRICT_TYPE_CHECKS
+#define NV_KUID_TO_UID(value) (__kuid_val(value))
+#else
+#define NV_KUID_TO_UID(value) (value)
+#endif
+
 #if defined(NV_TASK_STRUCT_HAS_CRED)
 #define NV_CURRENT_EUID() \
-    (((typeof(*current->cred) __force __kernel *)current->cred)->euid)
+    NV_KUID_TO_UID(((typeof(*current->cred) __force __kernel *)current->cred)->euid)
 #else
-#define NV_CURRENT_EUID() (current->euid)
+#define NV_CURRENT_EUID() NV_KUID_TO_UID(current->euid)
 #endif
 
 #if !defined(VM_FAULT_NOPAGE)
-- 
1.7.9.5

